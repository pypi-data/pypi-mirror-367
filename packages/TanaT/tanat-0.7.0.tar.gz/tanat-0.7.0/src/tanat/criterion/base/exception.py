#!/usr/bin/env python3
"""
Criterion exceptions.
"""

from ...exception import TanatException


class CriterionException(TanatException):
    """
    Base exception for all criterion-related errors in the TanaT framework.
    """


class InvalidCriterionError(CriterionException):
    """
    Raised when an invalid or improperly configured criterion is encountered.
    """
