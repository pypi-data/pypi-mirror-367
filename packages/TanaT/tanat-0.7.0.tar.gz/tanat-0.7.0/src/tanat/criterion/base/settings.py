#!/usr/bin/env python3
"""
Base class for criterion.
"""

import dataclasses

from .enum import CriterionLevel


@dataclasses.dataclass
class Criterion:
    """
    Base class for criterion.
    """

    _REGISTER_NAME = None

    def get_compatibility_levels(self, max_level=CriterionLevel.TRAJECTORY):
        """List the compatibility levels of the criterion.
        Compatibility levels are marked with decorators.

        Args:
            max_level (CriterionLevel, optional): The maximum level to consider.
                Defaults to Trajectory. Entity < Sequence < Trajectory

        Returns a list of CriterionLevel representing the compatibility levels of the
        criterion. The levels are "sequence","entity" or "trajectory".
        """
        if isinstance(max_level, str):
            max_level = CriterionLevel.from_str(max_level)

        compatibility_levels = getattr(self, "COMPATIBILITY_LEVELS", set())
        return [level for level in compatibility_levels if level <= max_level]
