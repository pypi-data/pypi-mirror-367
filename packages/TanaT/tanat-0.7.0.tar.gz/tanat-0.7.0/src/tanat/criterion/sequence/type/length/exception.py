#!/usr/bin/env python3
"""
Length criterion exception.
"""

from ....base.exception import CriterionException


class ContradictoryLengthCriterionError(CriterionException):
    """
    Raised when length-based filtering criterion are logically incompatible.

    This exception occurs during length criterion configuration when:
    - Attempting to use both 'gt' and 'ge' operators simultaneously
    - Attempting to use both 'lt' and 'le' operators simultaneously
    - Specifying lower and upper bounds that create an impossible condition
      (e.g., lower bound is greater than or equal to upper bound)

    Prevents silent failures in sequence length filtering by explicitly
    highlighting configuration errors.
    """
