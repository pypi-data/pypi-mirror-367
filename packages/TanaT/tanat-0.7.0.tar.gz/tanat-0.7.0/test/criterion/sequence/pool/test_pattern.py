#!/usr/bin/env python3
"""
Test pattern criterion applied to sequence pools.
"""

import pytest

from tanat.criterion.mixin.pattern.settings import PatternCriterion
from tanat.criterion.mixin.pattern.exception import InvalidColumnPatternError


class TestPatternCriterion:
    """
    Test pattern criterion applied to sequence pools.
    """

    @pytest.mark.parametrize(
        "pool_type,pattern",
        [
            ("event", {"event_type": "EMERGENCY"}),
            ("interval", {"medication": "ANTIBIOTIC"}),
        ],
    )
    def test_filter_entity_level(self, sequence_pools, pool_type, pattern, snapshot):
        """
        Event/Interval Sequence pool: Pattern criterion applied to entity level (inplace=False).
        """
        pool = sequence_pools[pool_type]
        original_data = pool.sequence_data.copy()

        filtered_pool = pool.filter(
            PatternCriterion(pattern=pattern), level="entity", inplace=False
        )
        assert pool.sequence_data.equals(original_data)
        assert filtered_pool is not pool
        assert filtered_pool.static_data.equals(
            pool.static_data
        )  # static data should be the same
        snapshot.assert_match(filtered_pool.sequence_data.to_csv())

    @pytest.mark.parametrize(
        "pool_type,pattern",
        [
            ("event", {"event_type": "EMERGENCY"}),
            ("interval", {"medication": "ANTIBIOTIC"}),
        ],
    )
    def test_filter_entity_level_inplace(
        self, sequence_pools, pool_type, pattern, snapshot
    ):
        """
        Event/Interval Sequence pool: Pattern criterion applied to entity level (inplace=True).
        """
        pool = sequence_pools[pool_type].copy()
        result = pool.filter(
            PatternCriterion(pattern=pattern), level="entity", inplace=True
        )
        assert result is None
        assert pool.static_data.equals(pool.static_data)
        snapshot.assert_match(pool.sequence_data.to_csv())

    @pytest.mark.parametrize(
        "pool_type,pattern",
        [
            ("event", {"event_type": "EMERGENCY"}),
            ("interval", {"medication": "ANTIBIOTIC"}),
        ],
    )
    def test_filter_entity_level_consistency(self, sequence_pools, pool_type, pattern):
        """
        Check filter entity level with inplace=True and inplace=False produce the same results.
        """
        original_pool = sequence_pools[pool_type]

        ## -- non inplace
        filtered_pool_non_inplace = original_pool.filter(
            PatternCriterion(pattern=pattern), level="entity", inplace=False
        )
        non_inplace_result = filtered_pool_non_inplace.sequence_data

        ## -- inplace
        pool_copy = original_pool.copy()
        pool_copy.filter(
            PatternCriterion(pattern=pattern), level="entity", inplace=True
        )
        inplace_result = pool_copy.sequence_data

        ## -- test
        assert non_inplace_result.equals(inplace_result)

    @pytest.mark.parametrize(
        "pool_type,pattern",
        [
            ("state", {"health_state": "TREATMENT"}),
        ],
    )
    def test_filter_state_entity_level_error(self, sequence_pools, pool_type, pattern):
        """
        State Sequence pool: Pattern criterion applied to entity level raises error.
        """
        pool = sequence_pools[pool_type]

        with pytest.raises(NotImplementedError):
            pool.filter(
                PatternCriterion(pattern=pattern), level="entity", inplace=False
            )

        with pytest.raises(NotImplementedError):
            pool.filter(PatternCriterion(pattern=pattern), level="entity", inplace=True)

    @pytest.mark.parametrize(
        "pool_type,pattern",
        [
            ("event", {"event_type": ["EMERGENCY", "CONSULTATION"]}),
            ("state", {"health_state": ["TREATMENT", "CONVALESCENCE"]}),
            ("interval", {"medication": ["ANTIBIOTIC", "PAIN_RELIEVER"]}),
        ],
    )
    def test_filter_sequence_level(self, sequence_pools, pool_type, pattern, snapshot):
        """
        Sequence pool: Pattern criterion applied to sequence level (inplace=False).
        """
        pool = sequence_pools[pool_type]
        original_data = pool.sequence_data.copy()
        criterion = PatternCriterion(pattern=pattern, contains=True)

        filtered_pool = pool.filter(criterion, level="sequence", inplace=False)
        assert pool.sequence_data.equals(original_data)
        assert filtered_pool is not pool
        snapshot.assert_match(filtered_pool.sequence_data.to_csv())
        snapshot.assert_match(filtered_pool.static_data.to_csv())

    @pytest.mark.parametrize(
        "pool_type,pattern",
        [
            ("event", {"event_type": ["EMERGENCY", "CONSULTATION"]}),
            ("state", {"health_state": ["TREATMENT", "CONVALESCENCE"]}),
            ("interval", {"medication": ["ANTIBIOTIC", "PAIN_RELIEVER"]}),
        ],
    )
    def test_filter_sequence_level_inplace(
        self, sequence_pools, pool_type, pattern, snapshot
    ):
        """
        Sequence pool: Pattern criterion applied to sequence level (inplace=True).
        """
        pool = sequence_pools[pool_type].copy()
        criterion = PatternCriterion(pattern=pattern, contains=True)
        result = pool.filter(criterion, level="sequence", inplace=True)
        assert result is None
        snapshot.assert_match(pool.sequence_data.to_csv())
        snapshot.assert_match(pool.static_data.to_csv())

    @pytest.mark.parametrize(
        "pool_type,level,pattern",
        [
            ("event", "entity", {"FAKE_COL": ["EMERGENCY", "CONSULTATION"]}),
            ("interval", "entity", {"FAKE_COL": ["ANTIBIOTIC", "PAIN_RELIEVER"]}),
            ("event", "sequence", {"FAKE_COL": ["EMERGENCY", "CONSULTATION"]}),
            ("state", "sequence", {"FAKE_COL": ["TREATMENT", "CONVALESCENCE"]}),
            ("interval", "sequence", {"FAKE_COL": ["ANTIBIOTIC", "PAIN_RELIEVER"]}),
        ],
    )
    def test_filter_raises_on_invalid_column(
        self, sequence_pools, pool_type, level, pattern
    ):
        """
        Ensure that filtering a sequence pool with an invalid column name
        in the pattern raises an InvalidColumnPatternError, both with
        inplace=False and inplace=True.
        """
        original_pool = sequence_pools[pool_type]
        criterion = PatternCriterion(pattern=pattern, contains=True)

        with pytest.raises(InvalidColumnPatternError):
            _ = original_pool.filter(criterion, level=level, inplace=False)

        with pytest.raises(InvalidColumnPatternError):
            pool_copy = original_pool.copy()
            pool_copy.filter(criterion, level=level, inplace=True)

    @pytest.mark.parametrize(
        "pool_type,pattern",
        [
            ("event", {"event_type": "FAKE_PATTERN"}),
            ("interval", {"medication": "FAKE_PATTERN"}),
        ],
    )
    def test_filter_no_matching_pattern_criterion_entity_level(
        self, sequence_pools, pool_type, pattern
    ):
        """
        Test filtering with pattern criterion at entity level that doesn't match any data.
        Should return an empty dataframe.
        """
        original_pool = sequence_pools[pool_type]
        criterion = PatternCriterion(pattern=pattern, contains=True)

        # -- non inplace
        filtered_pool_non_inplace = original_pool.filter(
            criterion, level="entity", inplace=False
        )
        ## -- test
        assert filtered_pool_non_inplace.sequence_data.empty
        assert filtered_pool_non_inplace.static_data.equals(original_pool.static_data)

        # -- inplace
        pool_copy = original_pool.copy()
        pool_copy.filter(criterion, level="entity", inplace=True)

        ## -- test
        assert pool_copy.sequence_data.empty
        assert pool_copy.static_data.equals(original_pool.static_data)

    @pytest.mark.parametrize(
        "pool_type,pattern",
        [
            ("event", {"event_type": "FAKE_PATTERN"}),
            ("state", {"health_state": "FAKE_PATTERN"}),
            ("interval", {"medication": "FAKE_PATTERN"}),
        ],
    )
    def test_filter_no_matching_pattern_criterion_sequence_level(
        self, sequence_pools, pool_type, pattern
    ):
        """
        Test filtering with pattern criterion at sequence level that doesn't match any data.
        Should return an empty dataframe.
        """
        original_pool = sequence_pools[pool_type]
        criterion = PatternCriterion(pattern=pattern, contains=True)

        # -- non inplace
        filtered_pool_non_inplace = original_pool.filter(
            criterion, level="sequence", inplace=False
        )
        ## -- test
        assert filtered_pool_non_inplace.sequence_data.empty
        assert filtered_pool_non_inplace.static_data.empty

        # -- inplace
        pool_copy = original_pool.copy()
        pool_copy.filter(criterion, level="sequence", inplace=True)

        ## -- test
        assert pool_copy.sequence_data.empty
        assert pool_copy.static_data.empty

    @pytest.mark.parametrize(
        "pool_type,pattern",
        [
            ("event", {"event_type": ["EMERGENCY", "CONSULTATION"]}),
            ("state", {"health_state": ["TREATMENT", "CONVALESCENCE"]}),
            ("interval", {"medication": ["ANTIBIOTIC", "PAIN_RELIEVER"]}),
        ],
    )
    def test_which_pattern_criterion(
        self, sequence_pools, pool_type, pattern, snapshot
    ):
        """Test which method for pattern criterion"""
        pool = sequence_pools[pool_type]
        pattern_criterion = PatternCriterion(pattern=pattern, contains=True)
        matching_ids = pool.which(pattern_criterion)
        assert isinstance(matching_ids, set)
        snapshot.assert_match(sorted(matching_ids))

    @pytest.mark.parametrize(
        "pool_type,pattern",
        [
            ("event", {"event_type": ["EMERGENCY", "CONSULTATION"]}),
            ("state", {"health_state": ["TREATMENT", "CONVALESCENCE"]}),
            ("interval", {"medication": ["ANTIBIOTIC", "PAIN_RELIEVER"]}),
        ],
    )
    def test_filter_sequence_level_consistency(
        self, sequence_pools, pool_type, pattern
    ):
        """
        Check consistency between filtering (inplace/non inplace) and which method.
        """
        original_pool = sequence_pools[pool_type]
        criterion = PatternCriterion(pattern=pattern, contains=True)

        # -- non inplace
        filtered_pool_non_inplace = original_pool.filter(
            criterion, level="sequence", inplace=False
        )
        non_inplace_result = filtered_pool_non_inplace.sequence_data

        # -- inplace
        pool_copy = original_pool.copy()
        pool_copy.filter(criterion, level="sequence", inplace=True)
        inplace_result = pool_copy.sequence_data

        ## -- which
        matching_ids = sorted(list(original_pool.which(criterion)))

        ## -- test
        assert non_inplace_result.equals(inplace_result)
        assert sorted(non_inplace_result.index.unique()) == matching_ids
        assert sorted(inplace_result.index.unique()) == matching_ids

    @pytest.mark.parametrize(
        "pool_type,pattern",
        [
            ("event", {"event_type": ["EMERGENCY", "CONSULTATION"]}),
            ("state", {"health_state": ["TREATMENT", "CONVALESCENCE"]}),
            ("interval", {"medication": ["ANTIBIOTIC", "PAIN_RELIEVER"]}),
        ],
    )
    def test_pattern_sequence_filtering_via_dict_criterion(
        self, sequence_pools, pool_type, pattern
    ):
        """
        Verify consistency of sequence filtering using criterion defined in a dictionary.
        """
        original_pool = sequence_pools[pool_type]
        criterion = {"pattern": pattern, "contains": True}

        # -- non inplace filtering
        filtered_pool_non_inplace = original_pool.filter(
            criterion=criterion,
            criterion_type="pattern",
            level="sequence",
            inplace=False,
        )
        non_inplace_result = filtered_pool_non_inplace.sequence_data

        # -- inplace filtering
        pool_copy = original_pool.copy()
        pool_copy.filter(
            criterion=criterion,
            criterion_type="pattern",
            level="sequence",
            inplace=True,
        )
        inplace_result = pool_copy.sequence_data

        # -- determine matching IDs
        matching_ids = sorted(
            original_pool.which(
                criterion=criterion,
                criterion_type="pattern",
            )
        )

        # -- assertions
        assert non_inplace_result.equals(inplace_result)
        assert sorted(non_inplace_result.index.unique()) == matching_ids
        assert sorted(inplace_result.index.unique()) == matching_ids
