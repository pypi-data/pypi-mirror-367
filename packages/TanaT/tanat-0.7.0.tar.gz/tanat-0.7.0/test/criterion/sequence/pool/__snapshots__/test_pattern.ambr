# serializer version: 1
# name: TestPatternCriterion.test_filter_entity_level[event-pattern0]
  '''
  patient_id,date,event_type,provider
  1,2023-08-03,EMERGENCY,NURSE
  1,2023-12-15,EMERGENCY,
  1,2023-12-26,EMERGENCY,
  2,2023-09-21,EMERGENCY,GENERAL_PRACTITIONER
  2,2023-09-28,EMERGENCY,RADIOLOGIST
  3,2023-03-22,EMERGENCY,
  3,2023-05-16,EMERGENCY,NURSE
  3,2023-06-21,EMERGENCY,NURSE
  3,2023-10-16,EMERGENCY,LABORATORY
  4,2023-06-09,EMERGENCY,NURSE
  4,2023-09-12,EMERGENCY,GENERAL_PRACTITIONER
  5,2023-03-22,EMERGENCY,
  5,2023-04-08,EMERGENCY,GENERAL_PRACTITIONER
  5,2023-06-01,EMERGENCY,GENERAL_PRACTITIONER
  5,2023-07-22,EMERGENCY,GENERAL_PRACTITIONER
  5,2023-08-21,EMERGENCY,NURSE
  6,2023-02-06,EMERGENCY,GENERAL_PRACTITIONER
  6,2023-05-04,EMERGENCY,LABORATORY
  6,2023-08-21,EMERGENCY,NURSE
  6,2023-12-12,EMERGENCY,NURSE
  8,2023-09-21,EMERGENCY,
  8,2023-11-13,EMERGENCY,
  10,2023-05-07,EMERGENCY,SPECIALIST
  10,2023-06-10,EMERGENCY,SPECIALIST
  
  '''
# ---
# name: TestPatternCriterion.test_filter_entity_level[interval-pattern1]
  '''
  patient_id,start_date,end_date,medication,administration_route,dosage
  1,2023-01-20,2023-02-13,ANTIBIOTIC,ORAL,2
  2,2023-01-05,2023-02-03,ANTIBIOTIC,ORAL,2
  2,2023-05-25,2023-06-09,ANTIBIOTIC,INHALATION,1
  2,2023-08-08,2023-08-19,ANTIBIOTIC,CUTANEOUS,3
  3,2023-04-15,2023-04-24,ANTIBIOTIC,INHALATION,3
  4,2023-05-25,2023-06-21,ANTIBIOTIC,ORAL,3
  5,2023-09-06,2023-09-18,ANTIBIOTIC,CUTANEOUS,2
  6,2023-05-18,2023-06-17,ANTIBIOTIC,INTRAMUSCULAR,1
  7,2023-03-14,2023-04-04,ANTIBIOTIC,,3
  8,2023-01-26,2023-02-05,ANTIBIOTIC,ORAL,3
  8,2023-01-28,2023-02-13,ANTIBIOTIC,ORAL,3
  9,2023-05-06,2023-05-24,ANTIBIOTIC,ORAL,1
  9,2023-09-21,2023-10-05,ANTIBIOTIC,CUTANEOUS,3
  10,2023-04-05,2023-05-05,ANTIBIOTIC,INTRAVENOUS,3
  
  '''
# ---
# name: TestPatternCriterion.test_filter_entity_level_inplace[event-pattern0]
  '''
  patient_id,date,event_type,provider
  1,2023-08-03,EMERGENCY,NURSE
  1,2023-12-15,EMERGENCY,
  1,2023-12-26,EMERGENCY,
  2,2023-09-21,EMERGENCY,GENERAL_PRACTITIONER
  2,2023-09-28,EMERGENCY,RADIOLOGIST
  3,2023-03-22,EMERGENCY,
  3,2023-05-16,EMERGENCY,NURSE
  3,2023-06-21,EMERGENCY,NURSE
  3,2023-10-16,EMERGENCY,LABORATORY
  4,2023-06-09,EMERGENCY,NURSE
  4,2023-09-12,EMERGENCY,GENERAL_PRACTITIONER
  5,2023-03-22,EMERGENCY,
  5,2023-04-08,EMERGENCY,GENERAL_PRACTITIONER
  5,2023-06-01,EMERGENCY,GENERAL_PRACTITIONER
  5,2023-07-22,EMERGENCY,GENERAL_PRACTITIONER
  5,2023-08-21,EMERGENCY,NURSE
  6,2023-02-06,EMERGENCY,GENERAL_PRACTITIONER
  6,2023-05-04,EMERGENCY,LABORATORY
  6,2023-08-21,EMERGENCY,NURSE
  6,2023-12-12,EMERGENCY,NURSE
  8,2023-09-21,EMERGENCY,
  8,2023-11-13,EMERGENCY,
  10,2023-05-07,EMERGENCY,SPECIALIST
  10,2023-06-10,EMERGENCY,SPECIALIST
  
  '''
# ---
# name: TestPatternCriterion.test_filter_entity_level_inplace[interval-pattern1]
  '''
  patient_id,start_date,end_date,medication,administration_route,dosage
  1,2023-01-20,2023-02-13,ANTIBIOTIC,ORAL,2
  2,2023-01-05,2023-02-03,ANTIBIOTIC,ORAL,2
  2,2023-05-25,2023-06-09,ANTIBIOTIC,INHALATION,1
  2,2023-08-08,2023-08-19,ANTIBIOTIC,CUTANEOUS,3
  3,2023-04-15,2023-04-24,ANTIBIOTIC,INHALATION,3
  4,2023-05-25,2023-06-21,ANTIBIOTIC,ORAL,3
  5,2023-09-06,2023-09-18,ANTIBIOTIC,CUTANEOUS,2
  6,2023-05-18,2023-06-17,ANTIBIOTIC,INTRAMUSCULAR,1
  7,2023-03-14,2023-04-04,ANTIBIOTIC,,3
  8,2023-01-26,2023-02-05,ANTIBIOTIC,ORAL,3
  8,2023-01-28,2023-02-13,ANTIBIOTIC,ORAL,3
  9,2023-05-06,2023-05-24,ANTIBIOTIC,ORAL,1
  9,2023-09-21,2023-10-05,ANTIBIOTIC,CUTANEOUS,3
  10,2023-04-05,2023-05-05,ANTIBIOTIC,INTRAVENOUS,3
  
  '''
# ---
# name: TestPatternCriterion.test_filter_sequence_level[event-pattern0]
  '''
  patient_id,date,event_type,provider
  3,2023-01-28,CONSULTATION,RADIOLOGIST
  3,2023-02-02,FOLLOW_UP,NURSE
  3,2023-02-10,FOLLOW_UP,LABORATORY
  3,2023-03-06,CONSULTATION,
  3,2023-03-22,EMERGENCY,
  3,2023-05-16,EMERGENCY,NURSE
  3,2023-06-21,EMERGENCY,NURSE
  3,2023-08-15,FOLLOW_UP,NURSE
  3,2023-08-19,EXAM,GENERAL_PRACTITIONER
  3,2023-09-09,VACCINATION,RADIOLOGIST
  3,2023-09-14,FOLLOW_UP,SPECIALIST
  3,2023-10-16,EMERGENCY,LABORATORY
  3,2023-12-10,CONSULTATION,RADIOLOGIST
  5,2023-01-02,EXAM,
  5,2023-02-21,VACCINATION,SPECIALIST
  5,2023-02-23,FOLLOW_UP,
  5,2023-03-22,EMERGENCY,
  5,2023-04-08,EMERGENCY,GENERAL_PRACTITIONER
  5,2023-04-23,FOLLOW_UP,GENERAL_PRACTITIONER
  5,2023-05-06,EXAM,LABORATORY
  5,2023-05-17,FOLLOW_UP,
  5,2023-06-01,EMERGENCY,GENERAL_PRACTITIONER
  5,2023-06-21,EXAM,RADIOLOGIST
  5,2023-06-29,CONSULTATION,GENERAL_PRACTITIONER
  5,2023-07-22,EMERGENCY,GENERAL_PRACTITIONER
  5,2023-08-19,CONSULTATION,GENERAL_PRACTITIONER
  5,2023-08-21,EMERGENCY,NURSE
  5,2023-09-12,FOLLOW_UP,RADIOLOGIST
  5,2023-10-10,CONSULTATION,
  5,2023-11-03,EXAM,RADIOLOGIST
  5,2023-11-27,EXAM,RADIOLOGIST
  6,2023-02-06,EMERGENCY,GENERAL_PRACTITIONER
  6,2023-03-01,CONSULTATION,SPECIALIST
  6,2023-04-22,CONSULTATION,RADIOLOGIST
  6,2023-05-04,EMERGENCY,LABORATORY
  6,2023-05-08,EXAM,
  6,2023-05-28,VACCINATION,GENERAL_PRACTITIONER
  6,2023-05-31,EXAM,
  6,2023-08-21,EMERGENCY,NURSE
  6,2023-09-16,VACCINATION,RADIOLOGIST
  6,2023-09-20,VACCINATION,SPECIALIST
  6,2023-09-24,VACCINATION,GENERAL_PRACTITIONER
  6,2023-10-31,VACCINATION,LABORATORY
  6,2023-12-12,EMERGENCY,NURSE
  
  '''
# ---
# name: TestPatternCriterion.test_filter_sequence_level[event-pattern0].1
  '''
  patient_id,gender,age,insurance,chronic_condition
  3,M,22,MIXED,False
  5,M,35,MIXED,False
  6,F,68,PUBLIC,False
  
  '''
# ---
# name: TestPatternCriterion.test_filter_sequence_level[interval-pattern2]
  '''
  patient_id,start_date,end_date,medication,administration_route,dosage
  2,2023-01-05,2023-02-03,ANTIBIOTIC,ORAL,2
  2,2023-01-10,2023-02-09,PAIN_RELIEVER,,1
  2,2023-02-04,2023-02-16,PAIN_RELIEVER,INTRAVENOUS,1
  2,2023-05-25,2023-06-09,ANTIBIOTIC,INHALATION,1
  2,2023-08-08,2023-08-19,ANTIBIOTIC,CUTANEOUS,3
  2,2023-08-08,2023-08-30,ANTIHYPERTENSIVE,INHALATION,2
  3,2023-04-15,2023-04-24,ANTIBIOTIC,INHALATION,3
  3,2023-09-24,2023-10-21,PAIN_RELIEVER,ORAL,2
  3,2023-10-10,2023-11-02,ANTIHYPERTENSIVE,ORAL,2
  5,2023-02-23,2023-03-13,ANTIHYPERTENSIVE,INTRAVENOUS,1
  5,2023-03-07,2023-03-24,CORTICOSTEROID,CUTANEOUS,2
  5,2023-04-22,2023-05-13,ANTIHYPERTENSIVE,INHALATION,2
  5,2023-09-04,2023-09-20,CORTICOSTEROID,,1
  5,2023-09-06,2023-09-18,ANTIBIOTIC,CUTANEOUS,2
  5,2023-10-21,2023-11-14,PAIN_RELIEVER,INTRAMUSCULAR,2
  8,2023-01-14,2023-02-13,CORTICOSTEROID,,3
  8,2023-01-26,2023-02-05,ANTIBIOTIC,ORAL,3
  8,2023-01-28,2023-02-13,ANTIBIOTIC,ORAL,3
  8,2023-03-19,2023-04-18,PAIN_RELIEVER,INHALATION,2
  8,2023-08-26,2023-09-12,ANTICOAGULANT,INHALATION,2
  8,2023-09-10,2023-10-06,PAIN_RELIEVER,CUTANEOUS,3
  8,2023-10-21,2023-10-30,CORTICOSTEROID,ORAL,3
  9,2023-04-19,2023-05-13,ANTICOAGULANT,INTRAMUSCULAR,3
  9,2023-05-06,2023-05-24,ANTIBIOTIC,ORAL,1
  9,2023-07-31,2023-08-10,PAIN_RELIEVER,INTRAVENOUS,1
  9,2023-09-21,2023-10-05,ANTIBIOTIC,CUTANEOUS,3
  
  '''
# ---
# name: TestPatternCriterion.test_filter_sequence_level[interval-pattern2].1
  '''
  patient_id,gender,age,insurance,chronic_condition
  2,M,31,PRIVATE,False
  3,M,22,MIXED,False
  5,M,35,MIXED,False
  8,M,83,PRIVATE,True
  9,F,68,MIXED,True
  
  '''
# ---
# name: TestPatternCriterion.test_filter_sequence_level[state-pattern1]
  '''
  patient_id,start_date,end_date,health_state,condition
  6,2023-01-01,2023-03-09,REMISSION,DIGESTIVE
  6,2023-03-09,2023-08-16,TREATMENT,RESPIRATORY
  6,2023-08-16,2023-12-14,TREATMENT,DIGESTIVE
  6,2023-12-14,2023-12-31,CONVALESCENCE,DIGESTIVE
  
  '''
# ---
# name: TestPatternCriterion.test_filter_sequence_level[state-pattern1].1
  '''
  patient_id,gender,age,insurance,chronic_condition
  6,F,68,PUBLIC,False
  
  '''
# ---
# name: TestPatternCriterion.test_filter_sequence_level_inplace[event-pattern0]
  '''
  patient_id,date,event_type,provider
  3,2023-01-28,CONSULTATION,RADIOLOGIST
  3,2023-02-02,FOLLOW_UP,NURSE
  3,2023-02-10,FOLLOW_UP,LABORATORY
  3,2023-03-06,CONSULTATION,
  3,2023-03-22,EMERGENCY,
  3,2023-05-16,EMERGENCY,NURSE
  3,2023-06-21,EMERGENCY,NURSE
  3,2023-08-15,FOLLOW_UP,NURSE
  3,2023-08-19,EXAM,GENERAL_PRACTITIONER
  3,2023-09-09,VACCINATION,RADIOLOGIST
  3,2023-09-14,FOLLOW_UP,SPECIALIST
  3,2023-10-16,EMERGENCY,LABORATORY
  3,2023-12-10,CONSULTATION,RADIOLOGIST
  5,2023-01-02,EXAM,
  5,2023-02-21,VACCINATION,SPECIALIST
  5,2023-02-23,FOLLOW_UP,
  5,2023-03-22,EMERGENCY,
  5,2023-04-08,EMERGENCY,GENERAL_PRACTITIONER
  5,2023-04-23,FOLLOW_UP,GENERAL_PRACTITIONER
  5,2023-05-06,EXAM,LABORATORY
  5,2023-05-17,FOLLOW_UP,
  5,2023-06-01,EMERGENCY,GENERAL_PRACTITIONER
  5,2023-06-21,EXAM,RADIOLOGIST
  5,2023-06-29,CONSULTATION,GENERAL_PRACTITIONER
  5,2023-07-22,EMERGENCY,GENERAL_PRACTITIONER
  5,2023-08-19,CONSULTATION,GENERAL_PRACTITIONER
  5,2023-08-21,EMERGENCY,NURSE
  5,2023-09-12,FOLLOW_UP,RADIOLOGIST
  5,2023-10-10,CONSULTATION,
  5,2023-11-03,EXAM,RADIOLOGIST
  5,2023-11-27,EXAM,RADIOLOGIST
  6,2023-02-06,EMERGENCY,GENERAL_PRACTITIONER
  6,2023-03-01,CONSULTATION,SPECIALIST
  6,2023-04-22,CONSULTATION,RADIOLOGIST
  6,2023-05-04,EMERGENCY,LABORATORY
  6,2023-05-08,EXAM,
  6,2023-05-28,VACCINATION,GENERAL_PRACTITIONER
  6,2023-05-31,EXAM,
  6,2023-08-21,EMERGENCY,NURSE
  6,2023-09-16,VACCINATION,RADIOLOGIST
  6,2023-09-20,VACCINATION,SPECIALIST
  6,2023-09-24,VACCINATION,GENERAL_PRACTITIONER
  6,2023-10-31,VACCINATION,LABORATORY
  6,2023-12-12,EMERGENCY,NURSE
  
  '''
# ---
# name: TestPatternCriterion.test_filter_sequence_level_inplace[event-pattern0].1
  '''
  patient_id,gender,age,insurance,chronic_condition
  3,M,22,MIXED,False
  5,M,35,MIXED,False
  6,F,68,PUBLIC,False
  
  '''
# ---
# name: TestPatternCriterion.test_filter_sequence_level_inplace[interval-pattern2]
  '''
  patient_id,start_date,end_date,medication,administration_route,dosage
  2,2023-01-05,2023-02-03,ANTIBIOTIC,ORAL,2
  2,2023-01-10,2023-02-09,PAIN_RELIEVER,,1
  2,2023-02-04,2023-02-16,PAIN_RELIEVER,INTRAVENOUS,1
  2,2023-05-25,2023-06-09,ANTIBIOTIC,INHALATION,1
  2,2023-08-08,2023-08-19,ANTIBIOTIC,CUTANEOUS,3
  2,2023-08-08,2023-08-30,ANTIHYPERTENSIVE,INHALATION,2
  3,2023-04-15,2023-04-24,ANTIBIOTIC,INHALATION,3
  3,2023-09-24,2023-10-21,PAIN_RELIEVER,ORAL,2
  3,2023-10-10,2023-11-02,ANTIHYPERTENSIVE,ORAL,2
  5,2023-02-23,2023-03-13,ANTIHYPERTENSIVE,INTRAVENOUS,1
  5,2023-03-07,2023-03-24,CORTICOSTEROID,CUTANEOUS,2
  5,2023-04-22,2023-05-13,ANTIHYPERTENSIVE,INHALATION,2
  5,2023-09-04,2023-09-20,CORTICOSTEROID,,1
  5,2023-09-06,2023-09-18,ANTIBIOTIC,CUTANEOUS,2
  5,2023-10-21,2023-11-14,PAIN_RELIEVER,INTRAMUSCULAR,2
  8,2023-01-14,2023-02-13,CORTICOSTEROID,,3
  8,2023-01-26,2023-02-05,ANTIBIOTIC,ORAL,3
  8,2023-01-28,2023-02-13,ANTIBIOTIC,ORAL,3
  8,2023-03-19,2023-04-18,PAIN_RELIEVER,INHALATION,2
  8,2023-08-26,2023-09-12,ANTICOAGULANT,INHALATION,2
  8,2023-09-10,2023-10-06,PAIN_RELIEVER,CUTANEOUS,3
  8,2023-10-21,2023-10-30,CORTICOSTEROID,ORAL,3
  9,2023-04-19,2023-05-13,ANTICOAGULANT,INTRAMUSCULAR,3
  9,2023-05-06,2023-05-24,ANTIBIOTIC,ORAL,1
  9,2023-07-31,2023-08-10,PAIN_RELIEVER,INTRAVENOUS,1
  9,2023-09-21,2023-10-05,ANTIBIOTIC,CUTANEOUS,3
  
  '''
# ---
# name: TestPatternCriterion.test_filter_sequence_level_inplace[interval-pattern2].1
  '''
  patient_id,gender,age,insurance,chronic_condition
  2,M,31,PRIVATE,False
  3,M,22,MIXED,False
  5,M,35,MIXED,False
  8,M,83,PRIVATE,True
  9,F,68,MIXED,True
  
  '''
# ---
# name: TestPatternCriterion.test_filter_sequence_level_inplace[state-pattern1]
  '''
  patient_id,start_date,end_date,health_state,condition
  6,2023-01-01,2023-03-09,REMISSION,DIGESTIVE
  6,2023-03-09,2023-08-16,TREATMENT,RESPIRATORY
  6,2023-08-16,2023-12-14,TREATMENT,DIGESTIVE
  6,2023-12-14,2023-12-31,CONVALESCENCE,DIGESTIVE
  
  '''
# ---
# name: TestPatternCriterion.test_filter_sequence_level_inplace[state-pattern1].1
  '''
  patient_id,gender,age,insurance,chronic_condition
  6,F,68,PUBLIC,False
  
  '''
# ---
# name: TestPatternCriterion.test_which_pattern_criterion[event-pattern0]
  list([
    3,
    5,
    6,
  ])
# ---
# name: TestPatternCriterion.test_which_pattern_criterion[interval-pattern2]
  list([
    2,
    3,
    5,
    8,
    9,
  ])
# ---
# name: TestPatternCriterion.test_which_pattern_criterion[state-pattern1]
  list([
    6,
  ])
# ---
